sim-lib/src/sim_node.rs

```diff
@@ -0,0 +1,1011 @@
+use triggered::{Listener, Trigger};
+
+use crate::{LightningError, LightningNode, NodeInfo, PaymentOutcome, PaymentResult};
+
+#[derive(Debug)]
+/// ForwardingError represents the various errors that we can run into when forwarding payments in a simulated network.
+/// Since we're not using real lightning nodes, these errors are not obfuscated and can be propagated to the sending
+/// node and used for analysis.
+pub enum ForwardingError {
+    /// Zero amount htlcs are in valid in the protocol.
+    ZeroAmountHtlc,
+    /// The outgoing channel id was not found in the network graph.
+    ChannelNotFound(u64),
+    /// The node pubkey provided was not associated with the channel in the network graph.
+    NodeNotFound(PublicKey),
+    /// The channel has already forwarded a HTLC with the payment hash provided (to be removed when MPP support is
+    /// added).
+    PaymentHashExists(PaymentHash),
+    /// A htlc with the payment hash provided could not be found to resolve.
+    PaymentHashNotFound(PaymentHash),
+    /// The forwarding node did not have sufficient outgoing balance to forward the htlc (htlc amount / balance).
+    InsufficientBalance(u64, u64),
+    /// The htlc forwarded is less than the channel's advertised minimum htlc amount (htlc amount / minimum).
+    LessThanMinimum(u64, u64),
+    /// The htlc forwarded is more than the chanenl's advertised maximum htlc amount (htlc amount / maximum).
+    MoreThanMaximum(u64, u64),
+    /// The channel has reached its maximum allowable number of htlcs in flight (total in flight / maximim).
+    ExceedsInFlightCount(u64, u64),
+    /// The forwarded htlc's amount would push the channel over its maximum allowable in flight total
+    /// (htlc amount / total in flight / maximum).
+    ExceedsInFlightTotal(u64, u64, u64),
+    /// The forwarded htlc's cltv expiry exceeds the maximum value used to express block heights in bitcoin.
+    ExpiryInSeconds(u32),
+    /// The forwarded htlc has insufficient cltv delta for the channel's minimum delta (cltv delta / minimum).
+    InsufficientCltvDelta(u32, u32),
+    /// The forwarded htlc has insufficient fee for the channel's policy (fee / base fee / prop fee / expected fee).
+    InsufficientFee(u64, u64, u64, u64),
+    /// Sanity check on channel balances failed (capacity / node_1 balance / node_2 balance).
+    SanityCheckFailed(u64, u64, u64),
+}
+
+impl ForwardingError {
+    /// Returns a boolean indicating whether failure to forward a htlc is a critical error that warrants shutdown.
+    fn is_critical(&self) -> bool {
+        matches!(
+            self,
+            ForwardingError::ZeroAmountHtlc
+                | ForwardingError::ChannelNotFound(_)
+                | ForwardingError::NodeNotFound(_)
+                | ForwardingError::PaymentHashExists(_)
+                | ForwardingError::PaymentHashNotFound(_)
+                | ForwardingError::SanityCheckFailed(_, _, _)
+        )
+    }
+}
+
+impl Display for ForwardingError {
```

Comment:
Why not derive `thiserror::Error`? It would reduce LOC

Reply:
Because I forgot that it existed, will update!